#!/usr/bin/env python
import glob
import os

if not os.path.exists("log"):
    os.makedirs("log")
	
configfile: "config.yaml"

FASTQDIR = config["resources"]["fastqdir"]
SV = config["resources"]["sv_detection"]
FLAIR = config["resources"]["flair"]
PH = config["resources"]["phasing"] 
Epi = config["resources"]["epigenetics"]


rule dna:
    input:
        expand("{FASTQDIR}/Q7_pass_reads.fastq.gz"),
        expand("{FASTQDIR}/Nanoplot")
        if Epi == 'run':
            expand("{FASTQDIR}/nanopolish_output")
        if SV == 'run':
            if ALIGNER == 'lra':
                expand("{FASTQDIR}/{sample}_lra/sv_calls/{sample}_cutesv_filtered.vcf.gz"),
                expand("{FASTQDIR}/{sample}_lra/alignment/{sample}_lra.bam"),
                expand("{FASTQDIR}/{sample}_lra/qc")
            if ALIGNER == 'win':
                expand("{FASTQDIR}/{sample}_win/sv_calls/{sample}_cutesv_filtered.vcf.gz"),
                expand("{FASTQDIR}/{sample}_win/alignment/{sample}_win.bam"),
                expand("{FASTQDIR}/{sample}_win/qc")
            if ALIGNER == 'both':
                expand("{FASTQDIR}/{sample}_lra/sv_calls/{sample}_cutesv_filtered.vcf.gz"),
                expand("{FASTQDIR}/{sample}_lra/alignment/{sample}_lra.bam"),
                expand("{FASTQDIR}/{sample}_lra/qc")
                expand("{FASTQDIR}/{sample}_win/sv_calls/{sample}_cutesv_filtered.vcf.gz"),
                expand("{FASTQDIR}/{sample}_win/alignment/{sample}_win.bam"),
                expand("{FASTQDIR}/{sample}_win/qc"),
                expand("{FASTQDIR}/{sample}/sv_calls/{sample}_cutesv_filtered.vcf.gz")
        if PH == 'run':
            expand("{FASTQDIR}/whatshap_output")

rule rna:
    input:
        expand("{FASTQDIR}/Q7_pass_reads.fastq.gz"),
        expand("{FASTQDIR}/Nanoplot"),
        expand("{FASTQDIR}/pychopper_output")
